hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -289,6 +296,94 @@ public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
         return true;
     }
 
+    @Override
+    public boolean saveToSnapshot() {
+        if (snapshotTraverser == null) {
+            List<Entry<?, ?>> snapshotList = new ArrayList<>();
+
+            for (Entry<Byte, Long> e : wmState.entrySet()) {
+                Long timestamp = e.getValue();
+                if (timestamp != OBJECT_2_LONG_MAP_MIN_VALUE) {
+                    snapshotList.add(
+                            entry(
+                                    broadcastKey(WM_STATE_KEY),
+                                    WatermarkStateValue.wmValue(e.getKey(), timestamp)
+                            )
+                    );
+                }
+            }
+
+            for (Entry<?, Long> e : lastReceivedWm.entrySet()) {
+                Long timestamp = e.getValue();
+                if (timestamp != OBJECT_2_LONG_MAP_MIN_VALUE) {
+                    snapshotList.add(
+                            entry(
+                                    broadcastKey(LAST_RECEIVED_WM_KEY),
+                                    WatermarkStateValue.wmValue((Byte) e.getKey(), timestamp)
+                            )
+                    );
+                }
+            }
+
+            Stream<Entry<?, ?>> leftBufferStream = buffer[0].content()
+                    .stream()
+                    .map(row -> entry(
+                            ObjectArrayKey.project(row, joinInfo.leftEquiJoinIndices()),
+                            BufferSnapshotValue.bufferValue(row, 0)
+                    ));
+
+            Stream<Entry<?, ?>> rightBufferStream = buffer[1].content()
```

Comment:
we do not save `unusedEventsTracker` state to snapshot. This may mean that items without matching row on the other side in outer join will be lost after snapshot/restore and we will not emit row extended with nulls.

Reply:
Outer join case will be covered in separate PR.

Reply:
design in #23141 says about saving watermarks for outer joins, but does not mention `unusedEventsTracker`. I think if we lose its state after snapshot+restore, we may produce incorrect outer join result (missing null-extended rows).

Reply:
Good catch, btw.